exec/plugin-commands-script-runners/src/run.ts

```diff
@@ -300,6 +304,55 @@ ${renderCommands(rootScripts)}`
   return output
 }
 
+export interface RunScriptOptions {
+  enablePrePostScripts: boolean
+}
+
+export const runScript: (opts: {
+  manifest: ProjectManifest
+  lifecycleOpts: RunLifecycleHookOptions
+  runScriptOptions: RunScriptOptions
+  passedThruArgs: string[]
+}, scriptName: string) => Promise<void> = async function (opts, scriptName) {
+  if (
+    opts.runScriptOptions.enablePrePostScripts &&
+    opts.manifest.scripts?.[`pre${scriptName}`] &&
+    !opts.manifest.scripts[scriptName].includes(`pre${scriptName}`)
+  ) {
+    await runLifecycleHook(`pre${scriptName}`, opts.manifest, opts.lifecycleOpts)
+  }
+  await runLifecycleHook(scriptName, opts.manifest, { ...opts.lifecycleOpts, args: opts.passedThruArgs })
+  if (
+    opts.runScriptOptions.enablePrePostScripts &&
+    opts.manifest.scripts?.[`post${scriptName}`] &&
+    !opts.manifest.scripts[scriptName].includes(`post${scriptName}`)
+  ) {
+    await runLifecycleHook(`post${scriptName}`, opts.manifest, opts.lifecycleOpts)
+  }
+}
+
 function renderCommands (commands: string[][]) {
   return commands.map(([scriptName, script]) => `  ${scriptName}\n    ${script}`).join('\n')
 }
+
+function getSpecifiedScripts (scripts: PackageScripts, scriptName: string) {
+  const scriptSelector = tryBuildRegExpFromCommand(scriptName)
+
+  // if scriptName which a user passes is RegExp (like /build:.*/), multiple scripts to execute will be selected with RegExp
+  if (scriptSelector) {
+    const scriptKeys = Object.keys(scripts)
+    return scriptKeys.filter(script => script.match(scriptSelector))
+  }
+
+  // if scripts in package.json has script which is equal to scriptName a user passes, return it.
+  if (scripts[scriptName]) {
+    return [scriptName]
+  }
+
+  // if a user passes start command as scriptName, `node server.js` will be executed as a fallback, so return start command even if start command is not defined in package.json
+  if (scriptName === 'start') {
+    return [scriptName]
+  }
+
+  return []
```

Comment:
We should do the complicated stuff last because in the vast majority of cases it is `scripts[scriptName]`.

```suggestion
  // if scripts in package.json has script which is equal to scriptName a user passes, return it.
  if (scripts[scriptName]) {
    return [scriptName]
  }

  // if a user passes start command as scriptName, `node server.js` will be executed as a fallback, so return start command even if start command is not defined in package.json
  if (scriptName === 'start') {
    return [scriptName]
  }

  // if scriptName which a user passes is RegExp (like /build:.*/), multiple scripts to execute will be selected with RegExp
  const scriptSelector = tryBuildRegExpFromCommand(scriptName)
  if (scriptSelector) {
    return Object.keys(scripts).filter(script => script.match(scriptSelector))
  }

  return []
```

Reply:
Indeed. [acf55fd](https://github.com/pnpm/pnpm/pull/5871/commits/acf55fd2ebe6c9fe9e9956b6690adc1d7ee40a3a)